lib/rspec/mocks/proxy.rb

```diff
@@ -406,34 +406,66 @@ class PartialClassDoubleProxy < PartialDoubleProxy
     # @private
     class ProxyForNil < PartialDoubleProxy
       def initialize(order_group)
-        @warn_about_expectations = true
+        set_expectation_behavior
         super(nil, order_group)
       end
 
+      attr_accessor :disallow_expectations
       attr_accessor :warn_about_expectations
       alias warn_about_expectations? warn_about_expectations
 
       def add_message_expectation(method_name, opts={}, &block)
-        warn(method_name) if warn_about_expectations?
+        warn_or_raise(method_name) if warn_or_disallow_expectations?
         super
       end
 
       def add_negative_message_expectation(location, method_name, &implementation)
-        warn(method_name) if warn_about_expectations?
+        warn_or_raise(method_name) if warn_or_disallow_expectations?
         super
       end
 
       def add_stub(method_name, opts={}, &implementation)
-        warn(method_name) if warn_about_expectations?
+        warn_or_raise(method_name) if warn_or_disallow_expectations?
         super
       end
 
     private
 
+      def set_expectation_behavior
+        case RSpec::Mocks.configuration.allow_message_expectations_on_nil
+        when false
+          @warn_about_expectations = false
+          @disallow_expectations = true
+        when true
+          @warn_about_expectations = false
+          @disallow_expectations = false
+        else
+          @warn_about_expectations = true
+          @disallow_expectations = false
+        end
+      end
+
+      def warn_or_disallow_expectations?
+        @disallow_expectations || @warn_about_expectations
+      end
+
+      def warn_or_raise(method_name)
+        if disallow_expectations
+          raise_error(method_name)
+        else
+          warn(method_name)
+        end
+      end
+
       def warn(method_name)
         source = CallerFilter.first_non_rspec_line
         Kernel.warn("An expectation of :#{method_name} was set on nil. Called from #{source}. Use allow_message_expectations_on_nil to disable warnings.")
       end
+
+      def raise_error(method_name)
+        source = CallerFilter.first_non_rspec_line
+        ErrorGenerator.new(self).raise_expectation_on_nil_error(method_name, source)
```

Comment:
It's a bit hard to tell in this context due to the subclassing but there's actually already an error generator instance available via `@error_generator` (see `Proxy#initialize`), so as I demonstrated above it's better to re-use that instance instead of making a new one.